### DISCUSSION TITLE
Advice for supporting Typescript 4.5 explicit eventual-extension-for-relative-imports policy?

### DISCUSSION THREAD

**Discussion by JasonKleban** [webpack/webpack#14478] [Q&A]
Please see https://github.com/microsoft/TypeScript/issues/46344, https://github.com/TypeStrong/ts-loader/issues/1383 for more details.

https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#type-in-packagejson-and-new-extensions

> This code works in CommonJS modules, but will fail in ES modules because relative import paths need to use extensions. As a result, it will have to be rewritten to use the extension of the output of foo.ts – so bar.ts will instead have to import from ./foo.js. ... This might feel a bit cumbersome at first, but TypeScript tooling like auto-imports and path completion will typically just do this for you.

[resolve.extensions](https://webpack.js.org/configuration/resolve/#resolveextensions) takes an array of suffixes, but doesn't seem to be able to express _replacements_ and doesn't seem to take such an expressive function, by the documentation.

Is there a way in current releases to support the Typescript 4.5 explicit eventual-extension-for-relative-imports policy?

**Comment by alexander-akait** [comment]
What is the problem with using `resolve.extensions`?

**Reply by alexander-akait** [reply]
Duplicate https://github.com/webpack/webpack/issues/13252?

**Reply by JasonKleban** [reply]
The typescript language service expects the extension and will give errors if there isn't one.  It complains about using the actual file extension because that would require a rewrite of a runtime portion of the typescript code.  resolve.extensions doesn't allow us to express such necessary replacement logic - and even if it did, that would probably be kinda tricky to get right.
